crutest/src/main.rs

```diff
@@ -744,9 +742,190 @@ async fn handle_signals(
     }
 }
 
+// Construct a volume for use by the tests.
+// Our choice of how to construct the volume depends on what options we
+// have been given.
+//
+// If we have been provided a vcr file, this will get first priority and all
+// other options will be ignored.
+//
+// Second choice is if we are provided the address for a dsc server.  We can
+// use the dsc server to determine part of what we need to create a Volume.
+// The rest of what we need we can gather from the CrucibleOpts, which are
+// built from options provided on the command line, or their defaults.
+//
+// For the final choice we have to construct a Volume by asking our downstairs
+// for information that we need up front, which we then combine with
+// CrucibleOpts.  This will work as long as one of the downstairs is up
+// already.  If we have a test that requires no downstairs to be running on
+// startup, then we need to provide a VCR file, or use the dsc server.
+async fn make_a_volume(
+    opt: &Opt,
+    block_io_logger: Logger,
+    test_log: &Logger,
+    pr: Option<ProducerRegistry>,
+) -> Result<Arc<Volume>> {
+    let up_uuid = opt.uuid.unwrap_or_else(Uuid::new_v4);
+    let mut crucible_opts = CrucibleOpts {
+        id: up_uuid,
+        target: opt.target.clone(),
+        lossy: opt.lossy,
+        flush_timeout: opt.flush_timeout,
+        key: opt.key.clone(),
+        cert_pem: opt.cert_pem.clone(),
+        key_pem: opt.key_pem.clone(),
+        root_cert_pem: opt.root_cert_pem.clone(),
+        control: opt.control,
+        read_only: opt.read_only,
+    };
+
+    if let Some(vcr_file) = &opt.vcr_file {
+        let vcr: VolumeConstructionRequest = match read_json(vcr_file) {
+            Ok(vcr) => vcr,
+            Err(e) => {
+                bail!("Error {:?} reading VCR from {:?}", e, vcr_file)
+            }
+        };
+        info!(test_log, "Using VCR: {:?}", vcr);
+
+        if opt.gen != 0 {
+            warn!(test_log, "gen option is ignored when VCR is provided");
+        }
+        if !opt.target.is_empty() {
+            warn!(test_log, "targets are ignored when VCR is provided");
+        }
+        let volume = Volume::construct(vcr, pr, block_io_logger).await.unwrap();
```

Comment:
Should we warn or error if `dsc` is provided here?

Personally, I would vote for errors (including for the lines that are warnings above), since invalid CLI arguments are probably a sign that the user is doing something wrong (and no one reads the logs).

Reply:
I think we are not quite ready to make these hard errors yet, though you are right that no one reads the logs, unless there is a failure.  This could lead to someone thinking they are testing one thing, when in fact they are testing another, which is not good.

My next PR (yes, I'm using that again) has more changes that will get us closer to not
needing both targets and a VCR file.  The 

There are some cases where we use the targets provided for building test specific args.  The `ReplaceBeforeActive`test you pointed to above is one of these cases.

I've made [an issue](https://github.com/oxidecomputer/crucible/issues/1473) to come back and harden up the errors here.

For `dsc` specifically, we could, in theory, create a volume from a VCR file, then use DSC to connect and control the downstairs themselves.  I'm trying to figure out how we will support having sub-volumes with different sizes and how to ingest that into crutest, and I was able to do that by layering a few dsc create commands in a way that did it.  
Another idea is to support multiple `dsc` endopints.  I'm still not sure which way is better.